vignettes/custom_epiworkflows.Rmd

```diff
@@ -0,0 +1,629 @@
+```
+</details>
+```{r, echo=FALSE}
+result_plot
+```
+
+## Population scaling
+
+Suppose we want to modify our predictions to apply to counts, rather than rates.
+To do that, we can adjust _just_ the `frosting` to perform post-processing on our existing rates forecaster.
+Since rates are calculated as counts per 100 000 people, we will convert back to counts by multiplying rates by the factor $\frac{regional \text{ } population}{100000}$.
+
+```{r rate_scale}
+count_layers <-
+  frosting() |>
+  layer_predict() |>
+  layer_residual_quantiles(quantile_levels = c(0.1, 0.25, 0.5, 0.75, 0.9)) |>
+  layer_population_scaling(
+    .pred,
+    .pred_distn,
+    # `df` contains scaling values for all regions; in this case it is the state populations
+    df = epidatasets::state_census,
+    df_pop_col = "pop",
+    create_new = FALSE,
+    # `rate_rescaling` gives the denominator of the existing rate predictions
+    rate_rescaling = 1e5,
+    by = c("geo_value" = "abbr")
+  ) |>
+  layer_add_forecast_date() |>
+  layer_add_target_date() |>
+  layer_threshold()
+
+# building the new workflow
+count_workflow <- epi_workflow(
+  four_week_recipe,
+  linear_reg(),
+  count_layers
+)
+count_pred_data <- get_test_data(four_week_recipe, training_data)
+count_predictions <- count_workflow |>
+  fit(training_data) |>
+  predict(count_pred_data)
+
+count_predictions
+```
+
+# Custom classifier workflow
+
+Let's work through an example of a more complicated kind of pipeline you can build using
+the `epipredict` framework.
+This is a hotspot prediction model, which predicts whether case rates are increasing (`up`), decreasing (`down`) or flat
+(`flat`).
+The model comes from a paper by McDonald, Bien, Green, Hu et al[^3], and roughly
+serves as an extension of `arx_classifier()`.
+
+First, we need to add a factor version of `geo_value`, so that it can be used as a feature.
+
+```{r training_factor}
+training_data <-
+  covid_case_death_rates |>
+  filter(time_value <= forecast_date, geo_value %in% used_locations) |>
+  mutate(geo_value_factor = as.factor(geo_value))
+```
+
+Then we put together the recipe, using a combination of base `{recipe}`
+functions such as `add_role()` and `step_dummy()`, and `{epipredict}` functions
+such as `step_growth_rate()`.
+
+```{r class_recipe}
+classifier_recipe <- epi_recipe(training_data) |>
+  # Label `time_value` as predictor and do no other processing
+  add_role(time_value, new_role = "predictor") |>
+  # Use one-hot encoding on `geo_value_factor` and label each resulting column as a predictor
+  step_dummy(geo_value_factor) |>
+  # Create and lag `case_rate` growth rate
+  step_growth_rate(case_rate, role = "none", prefix = "gr_") |>
+  step_epi_lag(starts_with("gr_"), lag = c(0, 7, 14)) |>
+  step_epi_ahead(starts_with("gr_"), ahead = 7, role = "none") |>
+  # Divide growth rate into 3 bins, and label as outcome variable
+  # Note `recipes::step_cut()` has a bug that prevents us from using it here
```

Comment:
This may be fixed. Can you check?

Reply:
It's kind of weird; if I just substitute it in, assigning roles via `step_cut(ahead_7_gr_7_rel_change_case_rate, breaks = c(-0.2, 0.25) / 7, role = "outcome")` doesn't actually work, but if I just let `step_epi_ahead` add the role, it works. `add_role` after doesn't seem to find the column for some mysterious reason. There's also no way to add labels to the values, so it's just predicts explicit ranges like `(0.0357, 0.262]` which I guess works.

Reply:
Looking at the code, it seems to always create closed ranges. I actually suspect this bug may have gone away, but I can't tell. Let's remove the note and add an issue to try to track it later.